test/regress_http.c

```diff
@@ -1304,6 +1307,206 @@ http_allowed_methods_test(void *arg)
+		test_ok++;
+	} else if (!strcmp(data, "Close")) {
+		evws_close(evws, 0);
+		test_ok++;
+	}
+}
+
+static void
+on_ws_close_cb(struct evws_connection *evws, void *arg)
+{
+	ev_uintptr_t val = (ev_uintptr_t)arg;
+
+	if (val != 0xDEADBEEF) {
+		fprintf(stdout, "FAILED on_complete_cb argument\n");
+		exit(1);
+	}
+	test_ok++;
+}
+
+static void
+http_on_ws_cb(struct evhttp_request *req, void *arg)
+{
+	struct evws_connection *evws;
+
+	evws = evws_new_session(req, on_ws_msg_cb, (void *)0xDEADBEEF);
+	if (!evws)
+		return;
+	test_ok++;
+
+	evws_connection_set_closecb(evws, on_ws_close_cb, (void *)0xDEADBEEF);
+	evws_send(evws, "Hello", strlen("Hello"));
+}
+
+static void
+http_ws_errorcb(struct bufferevent *bev, short what, void *arg)
+{
+	/** For ssl */
+	if (what & BEV_EVENT_CONNECTED)
+		return;
+	test_ok++;
+	event_base_loopexit(arg, NULL);
+}
+
+static void
+http_ws_readcb_phase2(struct bufferevent *bev, void *arg)
+{
+	event_base_loopexit(arg, NULL);
+}
+
+#define htonll(x) ((1==htonl(1)) ? (x) : \
+	((uint64_t)htonl((x) & 0xFFFFFFFF) << 32) | htonl((x) >> 32))
+#define ntohll(x) htonll(x)
+
+static void
+send_ws_msg(struct evbuffer* buf, const char* msg){
+	size_t len = strlen(msg);
+	uint8_t a = 0, b = 0, c = 0, d = 0;
+	uint8_t mask_key[4] = {1, 2, 3, 4}; /* should be random */
+	uint8_t m;
+
+	a |= 1 << 7;  /* fin */
+	a |= 1; /* text frame */
+
+	b |= 1 << 7; /* mask */
+
+	/* payload len */
+	if (len < 126) {
+		b |= len;
+	} else if (len < (1 << 16)) {
+		b |= 126;
+		c = htons(len);
+	} else {
+		b |= 127;
+		d = htonll(len);
+	}
+
+	evbuffer_add(buf, &a, 1);
+	evbuffer_add(buf, &b, 1);
+
+	if (c) evbuffer_add(buf, &c, sizeof(c));
+	else if (d) evbuffer_add(buf, &d, sizeof(d));
+
+	evbuffer_add(buf, &mask_key, 4);
+
+	for (size_t i = 0; i < len; i++) {
+		m = msg[i] ^ mask_key[i%4];
+		evbuffer_add(buf, &m, 1);
+	}
+}
+
+static void
+http_ws_readcb_hdr(struct bufferevent *bev, void *arg)
+{
+	struct evbuffer *input = bufferevent_get_input(bev);
+	struct evbuffer *output = bufferevent_get_input(bev);
+	size_t nread = 0, n = 0;
+	char* line;
+
+	while ((line = evbuffer_readln(input, &nread, EVBUFFER_EOL_CRLF))) {
+		if (n == 0 && !strncmp(line, "HTTP/1.1 101 ", strlen("HTTP/1.1 101 "))) {
+			test_ok++;
+		} else if (!strcmp(line, "Sec-WebSocket-Accept: HSmrc0sMlYUkAGmm5OPpG2HaGWk=")) {
+			test_ok++;
+		} else if (strlen(line) == 0) {
+			free(line);
+			bufferevent_setcb(bev, http_ws_readcb_phase2, http_writecb,
+					http_ws_errorcb, arg);
+			send_ws_msg(output, "Send echo");
+			test_ok++;
+			return;
+		}
+		free(line);
+		n++;
+	};
+}
+
+static void
+http_ws_readcb_bad(struct bufferevent *bev, void *arg)
+{
+	struct evbuffer *input = bufferevent_get_input(bev);
+	size_t nread;
+	char* line;
+
+	line = evbuffer_readln(input, &nread, EVBUFFER_EOL_CRLF);
+	if (!strncmp(line, "HTTP/1.1 401 ", strlen("HTTP/1.1 401 "))) {
+		test_ok++;
+	}
+	if (line)
+		free(line);
+}
+
+static void
+http_ws_test(void *arg)
```

Comment:
Now, you should register the test here - https://github.com/libevent/libevent/blob/bb57cea387ba4ebac077c71753e391225adfc81f/test/regress_http.c#L5899

Also consider moving tests into separate file, `test/regress_ws.c` ?